server/license: Enable the License Enforcer

spilchen opened the pull request:
Previously, the license enforcer was disabled by default and only enabled in specific test cases designed for it. This change fully activates the license enforcer and removes the option to disable it.

Some upgrade tests were failing because upgrades cannot be finalized without a valid license. To address this, I relaxed the restriction in test environments rather than modifying all mixed-mode tests to ensure a license is present.

We still do the enterprise check in this PR. Another PR (#131943) will be opened to remove them as it has quite a bit more test changes needed.

This change will be backported to versions 24.2, 24.1, 23.2, and 23.1.

Epic: CRDB-39988
Closes: CRDB-41758
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/131944)
<!-- Reviewable:end -->

fqazi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Just one minor question related to the environment variable

> Reviewed 9 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131944#-:-O8hBGjo9dudzLQ0KY_o:b47auk6)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @spilchen and @xinhaoz)

___
*[`pkg/upgrade/upgrades/v24_3_check_license_violation.go` line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/131944#-O8hB4x411I1zgVk7vIb:-O8hB4x411I1zgVk7vIc:b5diwpk) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cda5b47a8b484f52b6d04984f85a8fa7792ae71/pkg/upgrade/upgrades/v24_3_check_license_violation.go#L36)):*
> ```Go
> 	if deps.LicenseEnforcer.GetIsDisabled() ||
> 		(buildutil.CrdbTestBuild && (deps.TestingKnobs == nil || !deps.TestingKnobs.ForceCheckLicenseViolation)) ||
> 		envutil.EnvOrDefaultBool("COCKROACH_UPGRADE_SKIP_LICENSE_CHECK", false) {
> ```

Do we have tests that need this? I would think we would just set a license inside them


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131944)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @fqazi and @xinhaoz)

___
*[`pkg/upgrade/upgrades/v24_3_check_license_violation.go` line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/131944#-O8hB4x411I1zgVk7vIb:-O8hChKm7b7AL4cffIMv:b-booyjn) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cda5b47a8b484f52b6d04984f85a8fa7792ae71/pkg/upgrade/upgrades/v24_3_check_license_violation.go#L36)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Do we have tests that need this? I would think we would just set a license inside them
</blockquote></details>

This isn’t currently being used by anyone, but I thought it might be needed as a backdoor for mixed-version `roachprod` tests. Setting a license would also work around this issue, but I’m not sure how common that is in `roachprod` tests. I’m fine with removing it for now and only adding it back if the need arises. Let me know your thoughts.


<!-- Sent from Reviewable.io -->

spilchen committed:
server/license: Enable the License Enforcer

Previously, the license enforcer was disabled by default and only
enabled in specific test cases designed for it. This change fully
activates the license enforcer and removes the option to disable it.

Some upgrade tests were failing because upgrades cannot be finalized
without a valid license. To address this, I relaxed the restriction in
test environments rather than modifying all mixed-mode tests to ensure a
license is present.

This change will be backported to versions 24.2, 24.1, 23.2, and 23.1.

Epic: CRDB-39988
Closes: CRDB-41758
Release note: none

spilchen commented:
TFTR!

bors r=rafiss,fqazi
